connectorx-python/connectorx/__init__.py

```diff
@@ -247,6 +247,66 @@ def read_sql(
 ) -> pl.DataFrame: ...
 
 
+def get_passfile_content(path: pathlib.Path) -> dict:
+    # host:port:db_name:user_name:password
+    with open(path) as f:
+        contents = f.read().split(':')
+        if len(contents) != 5:
+            raise Exception('Pgpass content should follow: host:port:db_name:user_name:password')
+    return dict(zip(['hostname', 'port', 'path', 'username', 'password'], contents))
+
+
+def get_pgpass(conn) -> dict:
+    # check param or (PGPASS env or DEFAULT)
+    # DEFAULT = linux or windows
+    passfile = '%APPDATA%\postgresql\pgpass.conf' if sys.platform == 'windows' else os.path.expanduser(
+        '~/') + '.pgpass'
+    if 'passfile' in conn:
+        parsed_conn = urllib.parse.urlparse(conn)
+
+        # test if there is no '&'
+        for param in parsed_conn.params.split('&'):
+            k, v = param.split('=')
+            if k == 'passfile':
+                passfile = k
+
+    passfile_path = pathlib.Path(passfile)
+
+    if sys.platform != 'windows':
+        # In *nix platforms we check that the file is safe; it has to have 0600 permission.
+        # https://www.postgresql.org/docs/current/libpq-pgpass.html
+
+        try:
+            # We trust that the last three digits of st_mode are the permissions, e.g.
+            st_mode = int(oct(passfile_path.stat().st_mode)[-3:])
+
+        except Exception as e:
+            raise Exception(
+                'Could not check if file is safe, report to this to the maintainers please.') from e
+
+        if st_mode != 600:
+            raise Exception(
+                f'pgpass file does not have safe permissions (0600), it currently has "0{st_mode}" you can fix this buy running: $ chmod 0600 PASSFILE_PATH')
+
+    return get_passfile_content(passfile)
+
+
+def replace_conn_content(conn: str, contents: dict) -> str:
+    # We rewrite the netloc from scratch here e.g.
+    # netloc = contents['username'] or o.username + ':' + contents['password'] or o.password + ...
+    parsed_conn = urllib.parse.urlparse(conn)
+    return str(parsed_conn._replace(**contents))
```

Comment:
We can do the following:  
    database = parsed_conn.path[1:] or os.environ.get('PGDATABASE', os.getenv('USER'))
    host = parsed_conn.hostname or os.environ.get('PGHOST', 'localhost')
    port = parsed_conn.port or os.environ.get('PGPORT', '5432')
    user = parsed_conn.username or os.environ.get('PGUSER', os.getenv('USER'))
    password = parsed_conn.password

We parse 2 times the URI so maybe we need to rethink that maybe we call pgpass from inside reconstruct? That would allow something like that also:

query_params = parse_qs(parsed_con.query)

    passfile = query_params.get('passfile', [None])[0] or os.environ.get('PGPASSFILE')
    default_pgpass_path = os.path.expanduser('~/.pgpass') if os.name != 'nt' else os.path.join(os.getenv('APPDATA', ''), 'postgresql', 'pgpass.conf')

    # Determine the password precedence
    if not password:
        if 'PGPASSWORD' in os.environ:
            password = os.environ['PGPASSWORD']
        else:
            pgpass_path = passfile or default_pgpass_path
            password = get_passfile_content(pgpass_path, host or '*', str(port or '*'), database or '*', user or '*')

And with all these information we can get the new connection string.